FirebaseStorageUI/UIImageView+FirebaseStorage.m

```diff
@@ -0,0 +1,142 @@
+                                                                 FIRStorageReference *))completion{
+  return [self sd_setImageWithStorageReference:storageRef
+                                  maxImageSize:size
+                              placeholderImage:placeholder
+                                         cache:[SDImageCache sharedImageCache]
+                                    completion:completion];
+}
+
+- (id<FUIDownloadTask>)sd_setImageWithStorageReference:(id<FUIStorageReference>)storageRef
+                                          maxImageSize:(UInt64)size
+                                      placeholderImage:(nullable UIImage *)placeholder
+                                                 cache:(nullable id<FUIImageCache>)cache
+                                            completion:(void (^)(UIImage *,
+                                                                 NSError *,
+                                                                 SDImageCacheType,
+                                                                 FIRStorageReference *))completion {
+  NSParameterAssert(storageRef != nil);
+
+  // If there's already a download on this UIImageView, cancel it
+  if (self.sd_currentDownloadTask != nil) {
+    [self.sd_currentDownloadTask cancel];
+    self.sd_currentDownloadTask = nil;
+  }
+
+  // Set placeholder image
+  self.image = placeholder;
+
+  // Query cache for image before trying to download
+  NSString *key = storageRef.fullPath;
+  UIImage *cached = nil;
+
+  cached = [cache imageFromMemoryCacheForKey:key];
+  if (cached != nil) {
+    self.image = cached;
+    if (completion != nil) {
+      completion(cached, nil, SDImageCacheTypeMemory, storageRef);
+    }
+    return nil;
+  }
+
+  cached = [cache imageFromDiskCacheForKey:key];
+  if (cached != nil) {
+    self.image = cached;
+    if (completion != nil) {
+      completion(cached, nil, SDImageCacheTypeDisk, storageRef);
+    }
+    return nil;
+  }
+
+  // If nothing was found in cache, download the image from Firebase Storage
+  id<FUIDownloadTask> download = [storageRef dataWithMaxSize:size
+                                                  completion:^(NSData * _Nullable data,
+                                                               NSError * _Nullable error) {
+    dispatch_async(dispatch_get_main_queue(), ^{
+      if (data != nil) {
+        UIImage *image = [UIImage sd_imageWithData:data];
+        self.image = image;
+
+        // Cache downloaded image
+        [cache storeImage:image forKey:storageRef.fullPath];
+
+        if (completion != nil) {
+          completion(image, nil, SDImageCacheTypeNone, storageRef);
+        }
+      } else {
+        if (completion != nil) {
+          completion(nil, error, SDImageCacheTypeNone, storageRef);
+        }
+      }
+    });
+  }];
+  self.sd_currentDownloadTask = download;
+  return download;
+}
+
+#pragma mark - Accessors
```

Comment:
Might want to leave a comment here about why we're using associated objects, namely: http://stackoverflow.com/questions/12252725/defining-a-property-in-ios-class-extension